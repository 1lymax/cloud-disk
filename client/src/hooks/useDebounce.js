import {useCallback, useRef} from "react";


export const useDebounce = (callback, delay) => {
	const timer = useRef()

	return useCallback(() => {
		if (timer.current) {
			clearTimeout(timer.current)
		}
		timer.current = setTimeout(() => {
			callback()
		}, delay);

	}, [callback, delay]);


};