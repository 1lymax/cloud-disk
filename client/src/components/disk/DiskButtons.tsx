import {Button} from "@mui/material";
import React, {FC, useState} from 'react';
import SyncIcon from '@mui/icons-material/Sync';
import CreateNewFolderIcon from "@mui/icons-material/CreateNewFolder";
import KeyboardBackspaceIcon from "@mui/icons-material/KeyboardBackspace";

import ModalCreateDir from "../UI/ModalCreateDir";
import LoadingButton from "@mui/lab/LoadingButton";
import {popDirStack} from "../../store/reducers/FileSlice";
import {useAppDispatch, useAppSelector} from "../../hooks/hooks";


interface DiskButtonsProps {
    refetch: () => void;
    isLoading: boolean;
}


const DiskButtons: FC <DiskButtonsProps> = ({refetch, isLoading}) => {
    const dispatch = useAppDispatch()
    const [modalOpen, setModalOpen] = useState<boolean>(false);
    const currentDir = useAppSelector(state => state.fileState.currentDir)


    const handleBackButton = () => {
        dispatch(popDirStack())
    };

    return (
        <>
            {currentDir &&
				<Button size="large"
						startIcon={<KeyboardBackspaceIcon/>}
						variant="text"
						sx={{textTransform: "none"}}
						onClick={handleBackButton}
				>
					Back
				</Button>
            }
            <Button size="large"
                    startIcon={<CreateNewFolderIcon/>}
                    variant="text"
                    sx={{textTransform: "none"}}
                    onClick={() => setModalOpen(!modalOpen)}
            >
                Create Dir
            </Button>
            <LoadingButton loading={isLoading} size="large"
                    startIcon={<SyncIcon/>}
                    variant="text"
                    sx={{textTransform: "none"}}
                    onClick={() => refetch()}
            >
                Refresh
            </LoadingButton>
            <ModalCreateDir currentDir={currentDir} open={modalOpen} setOpen={setModalOpen}/>

        </>
    );
};

export default DiskButtons;