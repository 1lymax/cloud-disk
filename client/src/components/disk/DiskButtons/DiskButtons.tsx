import {Button} from "@mui/material";
import React, {FC, useState} from 'react';
import SyncIcon from '@mui/icons-material/Sync';
import CreateNewFolderIcon from "@mui/icons-material/CreateNewFolder";
import KeyboardBackspaceIcon from "@mui/icons-material/KeyboardBackspace";

import ModalCreateDir from "../../UI/ModalCreateDir";
import LoadingButton from "@mui/lab/LoadingButton";
import {popDirStack} from "../../../store/reducers/FileSlice";
import {useAppDispatch, useAppSelector} from "../../../hooks/hooks";


interface DiskButtonsProps {
    refetch: () => void;
    isLoading: boolean;
}


const DiskButtons: FC<DiskButtonsProps> = ({refetch, isLoading}) => {
    const dispatch = useAppDispatch()
    const [modalOpen, setModalOpen] = useState<boolean>(false);
    const currentDir = useAppSelector(state => state.fileState.currentDir)


    const handleBackButton = () => {
        dispatch(popDirStack())
    };

    return (
        <div data-testid='diskButtons'>
            {currentDir &&
				<Button size="large"
						data-testid='backDir-button'
						startIcon={<KeyboardBackspaceIcon/>}
						variant="text"
						sx={{textTransform: "none"}}
						onClick={handleBackButton}
				>
					Back
				</Button>
            }
            <Button size="large"
                    data-testid='createDir-button'
                    startIcon={<CreateNewFolderIcon/>}
                    variant="text"
                    sx={{textTransform: "none"}}
                    onClick={() => setModalOpen(!modalOpen)}
            >
                Create Dir
            </Button>
            <LoadingButton loading={isLoading} size="large"
                           startIcon={<SyncIcon/>}
                           data-testid='refresh-button'
                           variant="text"
                           sx={{textTransform: "none"}}
                           onClick={() => refetch()}
            >
                Refresh
            </LoadingButton>
            <ModalCreateDir currentDir={currentDir} open={modalOpen} setOpen={setModalOpen}/>

        </div>
    );
};

export default DiskButtons;