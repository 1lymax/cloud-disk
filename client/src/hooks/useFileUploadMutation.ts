import axios, {AxiosError} from 'axios'
import {Key} from 'react'
import {useMutation} from 'react-query'
import {LocalStorage} from "ts-localstorage";
import {ACCESS_TOKEN} from "../utils/consts";
import {useAppDispatch} from "./hooks";
import {updateUploadProgress} from "../store/reducers/FileSlice";

interface IArgs {
    file: File,
    parent: Key
}

export const useFileUploadMutation = () => {
    const dispatch = useAppDispatch()

    const mutation = useMutation<void, AxiosError, any>(
        args => axios.post(
            'http://localhost:5000/api/files/upload',
            args,
            {
                headers: {
                    'authorization': 'Bearer ' + LocalStorage.getItem(ACCESS_TOKEN),
                },
                onUploadProgress: ev =>
                    dispatch(updateUploadProgress({
                            file: args.get('file').name,
                            progress: Math.round((ev.loaded * 100) / (args.get('file').size))
                        })
                    )
            }
        )
    )

    return {...mutation}
}