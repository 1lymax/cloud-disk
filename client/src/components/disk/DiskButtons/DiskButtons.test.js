import {renderTestApp} from "../../../tests/helpers/renderTestApp";
import {screen, waitFor} from "@testing-library/react";
import DiskButtons from "./DiskButtons";
import {renderWithRedux} from "../../../tests/helpers/renderWithRedux";
import userEvent from "@testing-library/user-event";
import {useAppSelector} from "../../../hooks/hooks";

describe('DiskButtons', () => {
	test('match snapshot', async () => {
		renderWithRedux(<DiskButtons/>)
		expect(await screen.getByTestId('diskButtons')).toMatchSnapshot()

	})

	test('backDir button is present', async () => {
		renderWithRedux(<DiskButtons/>, {
			initialState: {
				fileState: {
					currentDir: '1231223'
				}
			}
		})
		expect(await screen.getByTestId('backDir-button')).toBeInTheDocument()

	})

	test('backDir button is present and pops dirStack state', async () => {
		renderWithRedux(<DiskButtons/>, {
			initialState: {
				fileState: {
					currentDir: '1231223',
					dirStack: ['123123']
				}
			}
		})
		const {dirStack} = useAppSelector(state => state.fileState)
		expect(dirStack).toHaveLength(1)
		const backDirButton = await screen.getByTestId('backDir-button')
		expect(backDirButton).toBeInTheDocument()
		await userEvent.click(backDirButton)
		expect(dirStack).toHaveLength(0)

	})

	test('createDir with modal open and close', async () => {
		renderWithRedux(<DiskButtons/>)
		screen.debug()
		const createDirButton = await screen.getByTestId('createDir-button')
		expect(createDirButton).toBeInTheDocument()

		// modal is closed
		expect(await screen.queryByTestId('modalCreateDir-dialog')).not.toBeInTheDocument()

		await userEvent.click(createDirButton)
		expect(await screen.getByTestId('modalCreateDir-dialog')).toBeInTheDocument()
		const createDirInput = await screen.getByTestId('createDir-input')
		const createDirSubmitButton = await screen.getByTestId('createDir-submit-button')

		expect(createDirInput).toBeInTheDocument()
		expect(createDirSubmitButton).toBeInTheDocument()
		await userEvent.type(createDirInput, '123')
		await userEvent.click(createDirSubmitButton)
		expect(createDirSubmitButton).toBeDisabled()

		await waitFor(() => {
			expect(screen.queryByTestId('modalCreateDir-dialog')).not.toBeInTheDocument()
		})
	})

	test('createDir with modal open and press cancel', async () => {
		renderWithRedux(<DiskButtons/>)
		screen.debug()
		const createDirButton = await screen.getByTestId('createDir-button')
		expect(createDirButton).toBeInTheDocument()

		// modal is closed
		expect(await screen.queryByTestId('modalCreateDir-dialog')).not.toBeInTheDocument()

		await userEvent.click(createDirButton)
		expect(await screen.getByTestId('modalCreateDir-dialog')).toBeInTheDocument()
		const createDirCancelButton = await screen.getByTestId('createDir-submit-button')

		expect(createDirCancelButton).toBeInTheDocument()
		await userEvent.click(createDirCancelButton)

		await waitFor(() => {
			expect(screen.queryByTestId('modalCreateDir-dialog')).not.toBeInTheDocument()
		})
	})

});
