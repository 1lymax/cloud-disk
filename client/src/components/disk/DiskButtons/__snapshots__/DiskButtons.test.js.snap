// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiskButtons match snapshot 1`] = `
<div
  data-testid="diskButtons"
>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeLarge MuiButton-textSizeLarge MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeLarge MuiButton-textSizeLarge css-1b4y6jr-MuiButtonBase-root-MuiButton-root"
    data-testid="createDir-button"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-startIcon MuiButton-iconSizeLarge css-htszrh-MuiButton-startIcon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="CreateNewFolderIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M20 6h-8l-2-2H4c-1.11 0-1.99.89-1.99 2L2 18c0 1.11.89 2 2 2h16c1.11 0 2-.89 2-2V8c0-1.11-.89-2-2-2zm-1 8h-3v3h-2v-3h-3v-2h3V9h2v3h3v2z"
        />
      </svg>
    </span>
    Create Dir
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
  <button
    class="MuiButtonBase-root MuiButton-root MuiLoadingButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeLarge MuiButton-textSizeLarge MuiButton-root MuiLoadingButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeLarge MuiButton-textSizeLarge css-1rmws6b-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
    data-testid="refresh-button"
    id=":r0:"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-startIcon MuiButton-iconSizeLarge css-htszrh-MuiButton-startIcon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="SyncIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 4V1L8 5l4 4V6c3.31 0 6 2.69 6 6 0 1.01-.25 1.97-.7 2.8l1.46 1.46C19.54 15.03 20 13.57 20 12c0-4.42-3.58-8-8-8zm0 14c-3.31 0-6-2.69-6-6 0-1.01.25-1.97.7-2.8L5.24 7.74C4.46 8.97 4 10.43 4 12c0 4.42 3.58 8 8 8v3l4-4-4-4v3z"
        />
      </svg>
    </span>
    Refresh
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;
